common/flusher/flusher.go

```diff
@@ -0,0 +1,275 @@
+		flushDuration          int
+		bufferCount            int
+		bufferItemCapacity     int
+		bufferNearItemCapacity int
+		flushNotifierChan      chan struct{}
+		logger                 log.Logger
+		shutdownChan           channel.ShutdownOnce
+		writer                 *ItemWriter[T]
+
+		sync.Mutex
+		flushTimer       *time.Timer
+		flushItemsBuffer []FlushItem[T]
+		fullBufferChan   chan []FlushItem[T]
+		freeBufferChan   chan []FlushItem[T]
+	}
+	FlushItem[T any] struct {
+		item T
+		fut  *future.FutureImpl[struct{}]
+	}
+	ItemWriter[T any] interface {
+		Write(flushItem ...FlushItem[T]) error
+	}
+	itemWriterImpl[T any] struct {
+		logger log.Logger
+		sync.Mutex
+	}
+)
+
+var _ ItemWriter = (*itemWriterImpl)(nil)
+var _ Flusher = (*flusherImpl)(nil)
+
+const (
+	bufferCapacityGapPercent = 0.9
+)
+
+func NewFlusher[T any](
+	bufferCount int,
+	bufferItemCapacity int,
+	flushDuration int,
+	writer *ItemWriter[T],
+	logger log.Logger,
+) *flusherImpl[T] {
+	return &flusherImpl[T]{
+		status:                 common.DaemonStatusInitialized,
+		flushDuration:          flushDuration,                                               // time waited after first item insertion before flushing the buffer
+		bufferCount:            bufferCount,                                                 // no of total flush buffers
+		bufferItemCapacity:     bufferItemCapacity,                                          // buffer size, will flush a buffer once no of items added to the buffer nears this limit
+		bufferNearItemCapacity: int(bufferCapacityGapPercent * float64(bufferItemCapacity)), // ^ will flush buffer once buffer size hits this number
+		flushTimer:             nil,
+		flushNotifierChan:      make(chan struct{}, 1),
+		writer:                 writer,
+		flushItemsBuffer:       make([]FlushItem[T], 0, bufferItemCapacity),
+		fullBufferChan:         make(chan []FlushItem[T], bufferCount),
+		freeBufferChan:         make(chan []FlushItem[T], bufferCount),
+		logger:                 logger,
+		shutdownChan:           channel.NewShutdownOnce(),
+	}
+}
+
+func NewItemWriter[T any](logger log.Logger) *itemWriterImpl[T] {
+	return &itemWriterImpl[T]{
+		logger: logger,
+	}
+}
+
+func (iwi *itemWriterImpl[T]) Write(flushItem ...FlushItem[T]) error {
+	return nil
+}
+
+func (f *flusherImpl[T]) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&f.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	f.flushTimer = time.NewTimer(time.Duration(f.flushDuration))
+	f.flushTimer.Stop() // Stop the timer after creation since we only want timer to start running upon first item insertion
+	f.initFreeBuffers(f.bufferCount, f.bufferItemCapacity)
+
+	go f.timerHandler()
+	go f.fullBufferChanHandler()
+}
+
+func (f *flusherImpl[T]) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&f.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	for _, flushItem := range f.flushItemsBuffer {
```

Comment:
Do you need to acquire lock to access flushItemsBuffer?

Reply:
That's a good question. Since we're calling Shutdown() right before we access flushItemsBuffer, that should kill all other Goroutines and prevent any new item to be added to the flushItemsBuffer. So I don't see a scenario where flushItemsBuffer will get accessed somewhere else once Shutdown() is called.

Reply:
E.g. The timer loop is in clearBufferLocked and someone call Stop()